### ISSUE TITLE
JS errors on checkout (bugsnagJS)

### ISSUE THREAD

**Issue by luisramos0** [openfoodfoundation/openfoodnetwork#5192]
## Description
Since we added bugsnagJS we are seeing two common JS errors on checkout that are happening to quite a lot of users:
![image](https://user-images.githubusercontent.com/1640378/78877032-25490d80-7a48-11ea-82b0-9f7d6982784e.png)
Just in UK 93 and 53 users respectively.

These errors do not necessarily mean the checkout is broken for all these users, but I think we need to investigate these errors as they can be causing checkout problems. 

## Expected Behavior
No JS errors related to checkout. 

## Actual Behaviour
JS errors.

## Steps to Reproduce
Unkonwn so far.

type error on bugsnag:
https://app.bugsnag.com/yaycode/open-food-network-js-uk/errors/5e847cf9751a150018bc4723?filters[event.since][0]=30d&filters[error.status][0]=open

UnhandledRejection on bugsnag:
https://app.bugsnag.com/yaycode/open-food-network-js-uk/errors/5e847f9022ed50001754372a?filters[event.since][0]=30d&filters[error.status][0]=open

## Severity
I'd put this as S2, it's a critical feature that may be broken.

## Your Environment
links to UK prod but it's happening on other servers as well.

## Possible Fix
In one case we are getting http 400 responses from the server so it should be possible to at least   add some more specific debug info to the error message to help future debugging.

**Comment by luisramos0** [comment]
So, unhandled rejections are false alarms, they are all user errors, real examples are:
![image](https://user-images.githubusercontent.com/1640378/78984035-4596cc00-7b1d-11ea-9ea1-f0700d5d67d3.png)
or
![image](https://user-images.githubusercontent.com/1640378/78984062-4f203400-7b1d-11ea-9270-24447f32e97f.png)

I will try to understand why this is generating a bugsnag error on the browser (it should only be flagged on teh server side).

This is the JS explanation of unhandledrejection:
"The unhandledrejection event is sent to the global scope of a script when a JavaScript Promise that has no rejection handler is rejected"

**Comment by luisramos0** [comment]
I will improve the checkout Javascript error handling code  in #5202

**Comment by Matt-Yorkley** [comment]
Are these cases covered in feature specs?

**Comment by luisramos0** [comment]
yes, checkout_spec tests a failed payment with bogus payment method. This test relies on this js code to pass.
Here spec/features/consumer/shopping/checkout_spec.rb:412

**Comment by luisramos0** [comment]
Marked this as prod test as we will have to monitor bugsnag after #5202 gets to production.
Moving to test ready and marking it as blocked until #5202 gets to production.

**Comment by luisramos0** [comment]
I hope I am using the labels and the zenhub column correctly :see_no_evil:

**Comment by luisramos0** [comment]
@RachL just shared a screenshot of a user experiencing #5080 on the checkout page.
That could mean that the bugsnag alerts of this issue are not related to the checkout JS code we are improving in #5202 but rather with, for example, stripe failing to load on the checkout page and breaking the login button...

**Comment by RachL** [comment]
A bit more info from this last user (but still no console sorry):
- he is an old customer of the hub. Was using firefox desktop for month with no issues
- he managed to login with his phone on chrome
- incognito mode on desktop didn't work as well

More info as soon as I get them....

**Comment by luisramos0** [comment]
@RachL your second screenshot where the button was displaying correctly was not on checkout so I'd say we should re-open #5080 because this issue here is for checkout errors.

- he is an old customer of the hub. Was using firefox desktop for month with no issues

This problem comes from some recent change in OFN or from a change in his Firefox (firefox version, firefox settings, firefox extensions)

- incognito mode on desktop didn't work as well

That's interesting. Usually browser extensions dont work in private mode.

**Comment by RachL** [comment]
@luisramos0 I'm sorry I've failed to move forward: the user is not responding anymore I won't have any console info. Is it worth to re-open #5080 ?
Also what's the status of this issue: How to test it in production?

**Comment by luisramos0** [comment]
I think this is a dev-test/prod-test :-)
We need to check bugsnag and see if these errors still occur: none since yesterday, let's wait a few more days and look at bugsnag again.

I think that if user reports keep coming after this week, we need to reopen 5080.

**Comment by luisramos0** [comment]
I think the situation has improved a lot, very few errors this week. I can only see two users with problems, new errors. they are seeing the snail. I thought the last PR handled a 404 but apparenly it doesnt, I can create a new issue for that if it keeps happening. 2 users in one week in UK is not a lot but maybe worth doing it because it's checkout.

I marked the errors as fixed on bugsnag, I'll keep this here until next week so we can check again.

**Comment by luisramos0** [comment]
There are very few JS errors on checkout :tada:
Since the 21st of April when the js improvement PR was deployed we have 3 JS errors:
- one "unknown error" that happen to only one user (edge case on the Edge browser?)
- one error where checkout returns 404 - affected 2 users
- a JS error mounting the Stripe HTML element (affecting 5 users since 21st April in UK) - this is happening on Mobile Safari, may be an edge case for that browser?

I am closing this issue as the main problem is solved.
We can open issues for these remaining issues. They will not be S2s with these number of occurrences!?
